############################################################################
**Rboretum**: *Analysis and Visualization of Phylogenies and their Support*
############################################################################
.. |br| raw:: html

   <br />

The **Rboretum** package for R provides a number of functions that allow users to:

1. Easily compare topologies across datasets with:

  - Easy tip label conversion
  - multiPhylo summary tools
  - Tree trimming function that prunes datasets down to a common set of taxa

2. Get summary statistics from multiple-sequence alignment data directly in R (via Biopython + *reticulate*), including:

  - Percent GC
  - Percent Missing Data
  - Percent Parsimony-Informative Data
  - Percent Invariant
 
3. Extract site-by-site allelic signal from MSA data, and convert it to node support information that can be tabulated and plotted onto trees.
|br|

4. Plot custom phylogenetic trees with a wide range of options including:

  - Branch/tip coloring
  - Plot resizing
  - The addition of informative node labels

==================
Table of contents
==================
.. toctree::
  :titlesonly:
  :maxdepth: 1
    
  content/installation
  content/loading
  content/rboretum-suite
  

=========
License
=========

**Rboretum** is freely available under a GNU Public License (Version 3).

================
Acknowledgments
================

**Rboretum** was developed with support from an NSF Postdoctoral Fellowship (Award Number: 1812201)
    
============
Bug Reports
============

If you wish to report bugs :-( or suggest features :-), please do so on the `GitHub Issues Page <https://github.com/BobLiterman/Rboretum/issues>`_
  