

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>getAlignmentSupport &mdash; Rboretum 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../_static/Rboretum_Logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="getSpeciesComposition" href="getSpeciesComposition.html" />
    <link rel="prev" title="getAlignmentSpecies" href="getAlignmentSpecies.html" />
    <link href="../../_static/RAL_Style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> Rboretum
          

          
            
            <img src="../../_static/Rboretum_Logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html"><strong>Installing Rboretum</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="../loading.html"><strong>Loading Rboretum</strong></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../rboretum-suite.html"><strong>Rboretum Functions</strong></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="getAlignmentComposition.html"><strong>getAlignmentComposition</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentBreakdown.html"><strong>getAlignmentBreakdown</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentPatterns.html"><strong>getAlignmentPatterns</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentSignal.html"><strong>getAlignmentSignal</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentSpecies.html"><strong>getAlignmentSpecies</strong></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><strong>getAlignmentSupport</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getSpeciesComposition.html"><strong>getSpeciesComposition</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../plotting_functions/tandemPlotter.html"><strong>tandemPlotter</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/checkTips.html"><strong>checkTips</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/extractNodeAges.html"><strong>extractNodeAges</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getSharedTaxa.html"><strong>getSharedTaxa</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeClades.html"><strong>getTreeClades</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeRoot.html"><strong>getTreeRoot</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeSplits.html"><strong>getTreeSplits</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getUniqueTopologies.html"><strong>getUniqueTopologies</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/readRooted.html"><strong>readRooted</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/summarizeMultiphylo.html"><strong>summarizeMultiphylo</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeDater.html"><strong>treeDater</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeNamer.html"><strong>treeNamer</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeTrimmer.html"><strong>treeTrimmer</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/convertLabels.html"><strong>convertLabels</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/semiSorter.html"><strong>semiSorter</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/semiVector.html"><strong>semiVector</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/tableCount.html"><strong>tableCount</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/vectorSemi.html"><strong>vectorSemi</strong></a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Rboretum</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../rboretum-suite.html"><strong>Rboretum Functions</strong></a> &raquo;</li>
        
      <li><strong>getAlignmentSupport</strong></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/content/alignment_functions/getAlignmentSupport.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getalignmentsupport">
<span id="id1"></span><h1><strong>getAlignmentSupport</strong><a class="headerlink" href="#getalignmentsupport" title="Permalink to this headline">¶</a></h1>
<p><strong>getAlignmentSupport</strong> queries the the parsimony-based signal from alignments profiled via <strong>getAlignmentSignal</strong>. Users can query support for:</p>
<blockquote>
<div><ul class="simple">
<li><p>A specific clade or set of clades</p></li>
<li><p>All clades in a phylo object</p></li>
<li><p>All clades from a multiPhylo of trees</p></li>
</ul>
</div></blockquote>
<p>Users can filter the output from <strong>getAlignmentSignal</strong> with options including:</p>
<blockquote>
<div><ul class="simple">
<li><p>Flexible handling of missing data</p></li>
<li><p>Handling of indels</p></li>
<li><p>Separating results by variation pattern (ie. all sites, all parsimony-informative, only biallelic, only triallelic, etc.)</p></li>
</ul>
</div></blockquote>
<p>If the output from <strong>getAlignmentSignal</strong> contains data from mutilple alignments, results can be separated or pooled via <em>separate_signal</em></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><strong>Note:</strong> Rboretum currently does not support handling of degenerate bases, and classifies them broadly as N/missing. Support for degenerate nucleotides is in progress and set for a future release.</p>
</div>
<div class="section" id="function-and-arguments">
<h2>Function and Arguments<a class="headerlink" href="#function-and-arguments" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">getAlignmentSupport</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">signal</span><span class="p">,</span><span class="n">tree</span><span class="p">,</span><span class="n">include_root</span><span class="p">,</span><span class="n">clade</span><span class="p">,</span><span class="n">dataset_name</span><span class="p">,</span><span class="n">max_missing</span><span class="p">,</span><span class="n">separate_signal</span><span class="p">,</span><span class="n">only_parsinf</span><span class="p">,</span><span class="n">include_gap</span><span class="p">,</span><span class="n">only_gap</span><span class="p">,</span><span class="n">only_biallelic</span><span class="p">,</span><span class="n">only_triallelic</span><span class="p">,</span><span class="n">only_quadallelic</span><span class="p">,</span><span class="n">only_pentallelic</span><span class="p">,</span><span class="n">return_integer</span><span class="p">,</span><span class="n">return_table</span><span class="p">,</span><span class="n">existing_support</span><span class="p">){</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 12%" />
<col style="width: 88%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>signal</strong></p></td>
<td><p>The output from <strong>getAlignmentSignal</strong>, run with species specified by <em>tree</em> or <em>clade</em></p></td>
</tr>
<tr class="row-odd"><td><p><strong>tree</strong></p></td>
<td><p>A rooted Phylo object, or a rooted multiPhylo where all trees share 3+ taxa</p></td>
</tr>
<tr class="row-even"><td><p><strong>include_root</strong></p></td>
<td><p>Include support about the root split [Default: TRUE]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>clade</strong></p></td>
<td><p>Query support for specific clades by providing them as semicolon-delimted characters (ie. c(‘Species_A;Species_B’,’Species_X;Species_Y’))</p></td>
</tr>
<tr class="row-even"><td><p><strong>dataset_name</strong></p></td>
<td><p>A label for the support column(s). [Default: Alignment name + m&lt;MISSING&gt;]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>max_missing</strong></p></td>
<td><p>Only record support from columns where no more than <em>max_missing</em> taxa have missing data (N or degenerate bases) [Default: All species - 3]</p></td>
</tr>
<tr class="row-even"><td><p><strong>separate_signal</strong></p></td>
<td><p>If TRUE and the output from <strong>getAlignmentSignal</strong> has 2+ alignments, create a separate support column for each alignment [Default: TRUE]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>only_parsinf</strong></p></td>
<td><p>Only query support from variable sites with no singletons [Default: FALSE]</p></td>
</tr>
<tr class="row-even"><td><p><strong>include_gap</strong></p></td>
<td><p>Query support from variable sites where gaps are present [Default: TRUE]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>only_gap</strong></p></td>
<td><p>Only query support from variable sites where gaps are present [Default: FALSE]</p></td>
</tr>
<tr class="row-even"><td><p><strong>only_biallelic</strong></p></td>
<td><p>Only query support from biallelic sites [Default: FALSE]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>only_triallelic</strong></p></td>
<td><p>Only query support from triallelic sites [Default: FALSE]</p></td>
</tr>
<tr class="row-even"><td><p><strong>only_quadallelic</strong></p></td>
<td><p>Only query support from quadallelic sites [Default: FALSE]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>only_pentallelic</strong></p></td>
<td><p>Only query support from pentallelic sites [Default: FALSE]</p></td>
</tr>
<tr class="row-even"><td><p><strong>return_integer</strong></p></td>
<td><p>Instead of a dataframe, return a named integer with values summed across alignments</p></td>
</tr>
<tr class="row-odd"><td><p><strong>return_table</strong></p></td>
<td><p>Instead of a dataframe, return the output from table() (or a list), where names are semicolon-delimted clades</p></td>
</tr>
<tr class="row-even"><td><p><strong>existing_support</strong></p></td>
<td><p>Output from <strong>getAlignmentSupport</strong> run with the same <em>tree</em> or <em>clade</em> argument</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="example-usage">
<h2>Example Usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Script: Rboretum/docs/content/Doc_Function_Scripts/getAlignmentSupport.R</span>

<span class="nf">library</span><span class="p">(</span><span class="n">Rboretum</span><span class="p">)</span>
<span class="nf">sourceRboretum</span><span class="p">()</span>

<span class="c1"># Read in a single phylogeny</span>
<span class="n">myTree</span> <span class="o">&lt;-</span> <span class="nf">readRooted</span><span class="p">(</span><span class="n">rb_tree1_path</span><span class="p">,</span><span class="n">root_taxa</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;Species_C&#39;</span><span class="p">,</span><span class="s">&#39;Species_H&#39;</span><span class="p">))</span>

<span class="c1"># Extract the signal from an alignment file</span>
<span class="n">mySignal</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSignal</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_align1_path</span><span class="p">,</span><span class="n">species_info</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">)</span>

<span class="c1"># Map the signal from a single alignment onto a single tree topology</span>
<span class="n">myTreeSupport</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignal</span><span class="p">,</span><span class="n">include_root</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Note the column name specifies how many taxa were allowed to have missing data (default: Total - 3)</span>
<span class="n">myTreeSupport</span>

                                                                                                                               <span class="n">Clade</span> <span class="n">Gene_1.phy_m12</span>
<span class="m">1</span>  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>
<span class="m">2</span>                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">13</span>
<span class="m">3</span>                                                                                  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">9</span>
<span class="m">4</span>                                                                                                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span>             <span class="m">16</span>
<span class="m">5</span>                                                                                            <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">8</span>
<span class="m">6</span>                                                                                                      <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">25</span>
<span class="m">7</span>                                                                                                                <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>
<span class="m">8</span>                                                                                                                <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_H</span>             <span class="m">17</span>
<span class="m">9</span>                                                                                  <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">26</span>
<span class="m">10</span>                                                                                                     <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">20</span>
<span class="m">11</span>                                                                                                               <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">11</span>
<span class="m">12</span>                                                                                                               <span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span>             <span class="m">15</span>
<span class="m">13</span>                                                                                                               <span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">3</span>

<span class="c1"># Map the signal from a single alignment onto a single tree topology, and only consider sites where one taxon has missing data. Add this to previous data.</span>
<span class="n">myTreeSupport_m1</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignal</span><span class="p">,</span><span class="n">include_root</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span><span class="n">max_missing</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span><span class="n">existing_support</span> <span class="o">=</span> <span class="n">myTreeSupport</span><span class="p">)</span>
<span class="n">myTreeSupport_m1</span>
                                                                                                                               <span class="n">Clade</span> <span class="n">Gene_1.phy_m12</span> <span class="n">Gene_1.phy_m1</span>
<span class="m">1</span>  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>            <span class="m">14</span>
<span class="m">2</span>                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">13</span>            <span class="m">13</span>
<span class="m">3</span>                                                                                  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">9</span>             <span class="m">9</span>
<span class="m">4</span>                                                                                                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span>             <span class="m">16</span>            <span class="m">16</span>
<span class="m">5</span>                                                                                            <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">8</span>             <span class="m">7</span>
<span class="m">6</span>                                                                                                      <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">25</span>            <span class="m">25</span>
<span class="m">7</span>                                                                                                                <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>            <span class="m">13</span>
<span class="m">8</span>                                                                                                                <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_H</span>             <span class="m">17</span>            <span class="m">16</span>
<span class="m">9</span>                                                                                  <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">26</span>            <span class="m">23</span>
<span class="m">10</span>                                                                                                     <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">20</span>            <span class="m">20</span>
<span class="m">11</span>                                                                                                               <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">11</span>            <span class="m">10</span>
<span class="m">12</span>                                                                                                               <span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span>             <span class="m">15</span>            <span class="m">14</span>
<span class="m">13</span>                                                                                                               <span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">3</span>             <span class="m">3</span>

<span class="c1"># Map the signal from a multiple alignments onto a single tree topology</span>
<span class="n">mySignals</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSignal</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_all_align</span><span class="p">,</span><span class="n">species_info</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">)</span>

<span class="c1"># Get tree support, separated by alignment. Treat gaps as missing data. Only include parsimony-informative sites</span>
<span class="n">myTreeSupport_Sep</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignals</span><span class="p">,</span><span class="n">include_root</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span><span class="n">include_gap</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">only_parsinf</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Get same tree support, summed across alignments</span>
<span class="n">myTreeSupport_Sum</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">myTree</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignals</span><span class="p">,</span><span class="n">include_root</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span><span class="n">separate_signal</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">include_gap</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">only_parsinf</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="n">myTreeSupport_Sep</span>
                                                                                                                               <span class="n">Clade</span> <span class="n">Gene_1.phy_m12</span> <span class="n">Gene_2.phy_m12</span> <span class="n">Gene_3.phy_m12</span> <span class="n">Gene_4.phy_m12</span> <span class="n">Gene_5.phy_m12</span>
<span class="m">1</span>  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>             <span class="m">35</span>              <span class="m">5</span>             <span class="m">39</span>             <span class="m">20</span>
<span class="m">2</span>                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">11</span>             <span class="m">27</span>              <span class="m">1</span>              <span class="m">0</span>              <span class="m">2</span>
<span class="m">3</span>                                                                                  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">8</span>             <span class="m">63</span>              <span class="m">7</span>              <span class="m">2</span>              <span class="m">0</span>
<span class="m">4</span>                                                                                                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span>             <span class="m">16</span>            <span class="m">115</span>             <span class="m">52</span>             <span class="m">19</span>             <span class="m">16</span>
<span class="m">5</span>                                                                                            <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>              <span class="m">8</span>             <span class="m">24</span>              <span class="m">9</span>              <span class="m">2</span>             <span class="m">15</span>
<span class="m">6</span>                                                                                                      <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">22</span>             <span class="m">25</span>             <span class="m">14</span>              <span class="m">2</span>             <span class="m">33</span>
<span class="m">7</span>                                                                                                                <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_O</span>             <span class="m">14</span>             <span class="m">44</span>             <span class="m">51</span>             <span class="m">50</span>              <span class="m">2</span>
<span class="m">8</span>                                                                                                                <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_H</span>             <span class="m">17</span>             <span class="m">72</span>             <span class="m">13</span>             <span class="m">76</span>             <span class="m">29</span>
<span class="m">9</span>                                                                                  <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">26</span>              <span class="m">6</span>             <span class="m">39</span>             <span class="m">49</span>              <span class="m">0</span>
<span class="m">10</span>                                                                                                     <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">19</span>             <span class="m">16</span>              <span class="m">7</span>             <span class="m">49</span>              <span class="m">0</span>
<span class="m">11</span>                                                                                                               <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_N</span>             <span class="m">10</span>             <span class="m">86</span>             <span class="m">10</span>             <span class="m">40</span>             <span class="m">14</span>
<span class="m">12</span>                                                                                                               <span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span>             <span class="m">15</span>             <span class="m">33</span>             <span class="m">32</span>             <span class="m">32</span>             <span class="m">26</span>
<span class="m">13</span>

<span class="n">myTreeSupport_Sum</span>

                                                                                                                               <span class="n">Clade</span> <span class="n">Total_m12</span>
<span class="m">1</span>  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span><span class="p">;</span><span class="n">Species_O</span>       <span class="m">113</span>
<span class="m">2</span>                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>        <span class="m">41</span>
<span class="m">3</span>                                                                                  <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_E</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>        <span class="m">80</span>
<span class="m">4</span>                                                                                                                <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span>       <span class="m">218</span>
<span class="m">5</span>                                                                                            <span class="n">Species_A</span><span class="p">;</span><span class="n">Species_F</span><span class="p">;</span><span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>        <span class="m">58</span>
<span class="m">6</span>                                                                                                      <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_D</span><span class="p">;</span><span class="n">Species_O</span>        <span class="m">96</span>
<span class="m">7</span>                                                                                                                <span class="n">Species_B</span><span class="p">;</span><span class="n">Species_O</span>       <span class="m">161</span>
<span class="m">8</span>                                                                                                                <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_H</span>       <span class="m">207</span>
<span class="m">9</span>                                                                                  <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span><span class="p">;</span><span class="n">Species_N</span>       <span class="m">120</span>
<span class="m">10</span>                                                                                                     <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_I</span><span class="p">;</span><span class="n">Species_N</span>        <span class="m">91</span>
<span class="m">11</span>                                                                                                               <span class="n">Species_G</span><span class="p">;</span><span class="n">Species_N</span>       <span class="m">160</span>
<span class="m">12</span>                                                                                                               <span class="n">Species_J</span><span class="p">;</span><span class="n">Species_M</span>       <span class="m">138</span>
<span class="m">13</span>                                                                                                               <span class="n">Species_K</span><span class="p">;</span><span class="n">Species_L</span>       <span class="m">183</span>

<span class="c1"># Get support for a specific</span>
<span class="n">myTreeSupport_Clade</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">clade</span><span class="o">=</span><span class="s">&#39;Species_C;Species_H&#39;</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignals</span><span class="p">)</span>
<span class="n">myTreeSupport_Clade</span>

                <span class="n">Clade</span> <span class="n">Gene_1.phy_m12</span> <span class="n">Gene_2.phy_m12</span> <span class="n">Gene_3.phy_m12</span> <span class="n">Gene_4.phy_m12</span> <span class="n">Gene_5.phy_m12</span>
<span class="m">1</span> <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_H</span>             <span class="m">17</span>             <span class="m">88</span>             <span class="m">17</span>             <span class="m">82</span>             <span class="m">34</span>

<span class="c1"># Get support for a specific clade not found in any trees</span>
<span class="n">myTreeSupport_Clade</span> <span class="o">&lt;-</span> <span class="nf">getAlignmentSupport</span><span class="p">(</span><span class="n">clade</span><span class="o">=</span><span class="s">&#39;Species_C;Species_F&#39;</span><span class="p">,</span><span class="n">signal</span> <span class="o">=</span> <span class="n">mySignals</span><span class="p">)</span>
<span class="n">myTreeSupport_Clade</span>

                <span class="n">Clade</span> <span class="n">Gene_1.phy_m12</span> <span class="n">Gene_2.phy_m12</span> <span class="n">Gene_3.phy_m12</span> <span class="n">Gene_4.phy_m12</span> <span class="n">Gene_5.phy_m12</span>
<span class="m">1</span> <span class="n">Species_C</span><span class="p">;</span><span class="n">Species_F</span>              <span class="m">0</span>              <span class="m">0</span>              <span class="m">0</span>              <span class="m">0</span>              <span class="m">0</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="getSpeciesComposition.html" class="btn btn-neutral float-right" title="getSpeciesComposition" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="getAlignmentSpecies.html" class="btn btn-neutral float-left" title="getAlignmentSpecies" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Robert Literman

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>