

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>getAlignmentBreakdown &mdash; Rboretum 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../_static/Rboretum_Logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="getAlignmentPatterns" href="getAlignmentPatterns.html" />
    <link rel="prev" title="getAlignmentComposition" href="getAlignmentComposition.html" />
    <link href="../../_static/RAL_Style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> Rboretum
          

          
            
            <img src="../../_static/Rboretum_Logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html"><strong>Installing Rboretum</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="../loading.html"><strong>Loading Rboretum</strong></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../rboretum-suite.html"><strong>Rboretum Functions</strong></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="getAlignmentComposition.html"><strong>getAlignmentComposition</strong></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><strong>getAlignmentBreakdown</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentPatterns.html"><strong>getAlignmentPatterns</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentSignal.html"><strong>getAlignmentSignal</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentSpecies.html"><strong>getAlignmentSpecies</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getAlignmentSupport.html"><strong>getAlignmentSupport</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="getSpeciesComposition.html"><strong>getSpeciesComposition</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../plotting_functions/tandemPlotter.html"><strong>tandemPlotter</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../plotting_functions/treePlotter.html"><strong>treePlotter</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/checkTips.html"><strong>checkTips</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/extractNodeAges.html"><strong>extractNodeAges</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getSharedTaxa.html"><strong>getSharedTaxa</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeClades.html"><strong>getTreeClades</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeRoot.html"><strong>getTreeRoot</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getTreeSplits.html"><strong>getTreeSplits</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/getUniqueTopologies.html"><strong>getUniqueTopologies</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/readRooted.html"><strong>readRooted</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/summarizeMultiphylo.html"><strong>summarizeMultiphylo</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeDater.html"><strong>treeDater</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeNamer.html"><strong>treeNamer</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../tree_functions/treeTrimmer.html"><strong>treeTrimmer</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/convertLabels.html"><strong>convertLabels</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/semiSorter.html"><strong>semiSorter</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/semiVector.html"><strong>semiVector</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/tableCount.html"><strong>tableCount</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="../utility_functions/vectorSemi.html"><strong>vectorSemi</strong></a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Rboretum</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../rboretum-suite.html"><strong>Rboretum Functions</strong></a> &raquo;</li>
        
      <li><strong>getAlignmentBreakdown</strong></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/content/alignment_functions/getAlignmentBreakdown.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getalignmentbreakdown">
<span id="id1"></span><h1><strong>getAlignmentBreakdown</strong><a class="headerlink" href="#getalignmentbreakdown" title="Permalink to this headline">¶</a></h1>
<p><strong>getAlignmentBreakdown</strong> returns the species IDs from an alignment file, specified by <em>alignment_path</em></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><strong>Note:</strong> Rboretum currently does not support handling of degenerate bases, and classifies them broadly as N/missing. Support for degenerate nucleotides is in progress and set for a future release.</p>
</div>
<div class="section" id="function-and-arguments">
<h2>Function and Arguments<a class="headerlink" href="#function-and-arguments" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">getAlignmentBreakdown</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">alignment_path</span><span class="p">,</span><span class="n">species_info</span><span class="p">,</span><span class="n">use_gaps</span><span class="p">,</span><span class="n">alignment_name</span><span class="p">,</span><span class="n">prefix</span><span class="p">,</span><span class="n">suffix</span><span class="p">)</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 12%" />
<col style="width: 88%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>alignment_path</strong></p></td>
<td><p>An absolute or relative path to an alignment file(s) or a directory containing multiple alignments</p></td>
</tr>
<tr class="row-odd"><td><p><strong>species_info</strong></p></td>
<td><p>A list of species to consider, provided as a phylo, multiPhlyo, character vector or labels or a semicolon-delimited string [Default: Shared species among all alignments]</p></td>
</tr>
<tr class="row-even"><td><p><strong>use_gaps</strong></p></td>
<td><p>If FALSE, consider gaps (-) in alignments as missing data. [Default: TRUE, treat gaps as indel characters]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>alignment_name</strong></p></td>
<td><p>A character vector containing desired alignment IDs [Default: Derive name from filename]</p></td>
</tr>
<tr class="row-even"><td><p><strong>prefix</strong></p></td>
<td><p>If <strong>alignment_path</strong> points to a directory, only query files starting with <strong>prefix</strong> (e.g. ‘Alignment’) [Default: Use all files in directory]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>suffix</strong></p></td>
<td><p>If <strong>alignment_path</strong> points to a directory, only query files ending wtih <strong>suffix</strong> (e.g. ‘.nex’) [Default: Use all files in directory]</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="example-usage">
<h2>Example Usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Script: Rboretum/docs/content/Doc_Function_Scripts/getAlignmentBreakdown.R</span>

<span class="nf">library</span><span class="p">(</span><span class="n">Rboretum</span><span class="p">)</span>
<span class="nf">sourceRboretum</span><span class="p">()</span>

<span class="c1"># Get alignment Breakdown information for a single alignment</span>
<span class="nf">getAlignmentBreakdown</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_align1_path</span><span class="p">)</span>

  <span class="n">Alignment_Name</span> <span class="n">Alignment_Length</span> <span class="n">Alignment_Percent_GC</span> <span class="n">Percent_Degenerate</span>  <span class="n">Percent_N</span> <span class="n">Percent_Gap</span> <span class="n">Species_GC_Mean</span> <span class="n">Species_GC_StdDev</span> <span class="n">Percent_Nonbase</span> <span class="n">Percent_Invariant</span> <span class="n">Percent_Singleton</span> <span class="n">Percent_Parsimony_Informative</span>
<span class="m">1</span>     <span class="n">Gene_1.phy</span>             <span class="m">1551</span>            <span class="m">0.3982911</span>                  <span class="m">0</span> <span class="m">0.02677842</span>  <span class="m">0.03249516</span>       <span class="m">0.3982685</span>       <span class="m">0.006593726</span>               <span class="m">0</span>         <span class="m">0.7176015</span>         <span class="m">0.1476467</span>                     <span class="m">0.1263701</span>
  <span class="n">Percent_Biallelic</span> <span class="n">Percent_Triallelic</span> <span class="n">Percent_Quadallelic</span> <span class="n">Percent_Pentallelic</span>
<span class="m">1</span>         <span class="m">0.1289491</span>        <span class="m">0.005802708</span>                   <span class="m">0</span>                   <span class="m">0</span>

<span class="c1"># Get alignment Breakdown information from all .phy files in a directory, providing new names, consider gaps as missing data</span>
<span class="nf">getAlignmentBreakdown</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_alignment_dir</span><span class="p">,</span><span class="n">use_gaps</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">suffix</span> <span class="o">=</span> <span class="s">&quot;.phy&quot;</span><span class="p">,</span><span class="n">alignment_name</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;Gene_A&#39;</span><span class="p">,</span><span class="s">&#39;Gene_B&#39;</span><span class="p">,</span><span class="s">&#39;Gene_C&#39;</span><span class="p">,</span><span class="s">&#39;Gene_D&#39;</span><span class="p">,</span><span class="s">&#39;Gene_E&#39;</span><span class="p">))</span>

  <span class="n">Alignment_Name</span> <span class="n">Alignment_Length</span> <span class="n">Alignment_Percent_GC</span> <span class="n">Percent_Degenerate</span>  <span class="n">Percent_N</span> <span class="n">Percent_Gap</span> <span class="n">Species_GC_Mean</span> <span class="n">Species_GC_StdDev</span> <span class="n">Percent_Nonbase</span> <span class="n">Percent_Invariant</span> <span class="n">Percent_Singleton</span> <span class="n">Percent_Parsimony_Informative</span>
<span class="m">1</span>         <span class="n">Gene_A</span>             <span class="m">1551</span>            <span class="m">0.3982911</span>                  <span class="m">0</span> <span class="m">0.02677842</span>  <span class="m">0.03249516</span>       <span class="m">0.3982685</span>       <span class="m">0.006593726</span>      <span class="m">0.02707930</span>         <span class="m">0.7504836</span>         <span class="m">0.1121857</span>                     <span class="m">0.1044487</span>
<span class="m">2</span>         <span class="n">Gene_B</span>             <span class="m">2804</span>            <span class="m">0.5219202</span>                  <span class="m">0</span> <span class="m">0.02691393</span>  <span class="m">0.10537328</span>       <span class="m">0.5219258</span>       <span class="m">0.002919946</span>      <span class="m">0.09486448</span>         <span class="m">0.4925107</span>         <span class="m">0.2014979</span>                     <span class="m">0.1690442</span>
<span class="m">3</span>         <span class="n">Gene_C</span>             <span class="m">1031</span>            <span class="m">0.5919662</span>                  <span class="m">0</span> <span class="m">0.02644682</span>  <span class="m">0.27009376</span>       <span class="m">0.5919771</span>       <span class="m">0.009096305</span>      <span class="m">0.23957323</span>         <span class="m">0.2580019</span>         <span class="m">0.2085354</span>                     <span class="m">0.2279340</span>
<span class="m">4</span>         <span class="n">Gene_D</span>             <span class="m">2219</span>            <span class="m">0.4160601</span>                  <span class="m">0</span> <span class="m">0.02679886</span>  <span class="m">0.09397627</span>       <span class="m">0.4160605</span>       <span class="m">0.005487643</span>      <span class="m">0.07706174</span>         <span class="m">0.4173051</span>         <span class="m">0.2158630</span>                     <span class="m">0.2532672</span>
<span class="m">5</span>         <span class="n">Gene_E</span>             <span class="m">1500</span>            <span class="m">0.4895705</span>                  <span class="m">0</span> <span class="m">0.02626667</span>  <span class="m">0.00000000</span>       <span class="m">0.4895817</span>       <span class="m">0.006269950</span>      <span class="m">0.00000000</span>         <span class="m">0.5713333</span>         <span class="m">0.1900000</span>                     <span class="m">0.2006667</span>
  <span class="n">Percent_Biallelic</span> <span class="n">Percent_Triallelic</span> <span class="n">Percent_Quadallelic</span> <span class="n">Percent_Pentallelic</span>
<span class="m">1</span>         <span class="m">0.1057382</span>        <span class="m">0.004513217</span>        <span class="m">0.0000000000</span>                   <span class="m">0</span>
<span class="m">2</span>         <span class="m">0.1990014</span>        <span class="m">0.012125535</span>        <span class="m">0.0000000000</span>                   <span class="m">0</span>
<span class="m">3</span>         <span class="m">0.2638215</span>        <span class="m">0.030067895</span>        <span class="m">0.0000000000</span>                   <span class="m">0</span>
<span class="m">4</span>         <span class="m">0.2685895</span>        <span class="m">0.020730059</span>        <span class="m">0.0004506534</span>                   <span class="m">0</span>
<span class="m">5</span>         <span class="m">0.2240000</span>        <span class="m">0.014666667</span>        <span class="m">0.0000000000</span>                   <span class="m">0</span>

<span class="c1"># Get alignment Breakdown information from all .phy files in a directory, providing new names, consider gaps as indel data, and only analyze species A-E</span>
<span class="nf">getAlignmentBreakdown</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_alignment_dir</span><span class="p">,</span><span class="n">species_info</span> <span class="o">=</span> <span class="s">&#39;Species_A;Species_B;Species_C;Species_D;Species_E&#39;</span><span class="p">,</span><span class="n">use_gaps</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span><span class="n">suffix</span> <span class="o">=</span> <span class="s">&quot;.phy&quot;</span><span class="p">,</span><span class="n">alignment_name</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;Gene_A&#39;</span><span class="p">,</span><span class="s">&#39;Gene_B&#39;</span><span class="p">,</span><span class="s">&#39;Gene_C&#39;</span><span class="p">,</span><span class="s">&#39;Gene_D&#39;</span><span class="p">,</span><span class="s">&#39;Gene_E&#39;</span><span class="p">))</span>

<span class="c1"># Get alignment Breakdown from dummy alignment, with and without gap support</span>
<span class="nf">getAlignmentBreakdown</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_dummy_align_path</span><span class="p">)</span>

      <span class="n">Alignment_Name</span> <span class="n">Alignment_Length</span> <span class="n">Alignment_Percent_GC</span> <span class="n">Percent_Degenerate</span> <span class="n">Percent_N</span> <span class="n">Percent_Gap</span> <span class="n">Species_GC_Mean</span> <span class="n">Species_GC_StdDev</span> <span class="n">Percent_Nonbase</span> <span class="n">Percent_Invariant</span> <span class="n">Percent_Singleton</span> <span class="n">Percent_Parsimony_Informative</span>
<span class="m">1</span> <span class="n">Dummy_Alignment.fa</span>               <span class="m">13</span>            <span class="m">0.1470588</span>         <span class="m">0.03846154</span> <span class="m">0.1384615</span>  <span class="m">0.03846154</span>       <span class="m">0.1686147</span>         <span class="m">0.1825288</span>       <span class="m">0.1538462</span>         <span class="m">0.1538462</span>         <span class="m">0.2307692</span>                     <span class="m">0.3076923</span>
  <span class="n">Percent_Biallelic</span> <span class="n">Percent_Triallelic</span> <span class="n">Percent_Quadallelic</span> <span class="n">Percent_Pentallelic</span>
<span class="m">1</span>         <span class="m">0.2307692</span>         <span class="m">0.07692308</span>          <span class="m">0.07692308</span>          <span class="m">0.07692308</span>

<span class="nf">getAlignmentBreakdown</span><span class="p">(</span><span class="n">alignment_path</span> <span class="o">=</span> <span class="n">rb_dummy_align_path</span><span class="p">,</span><span class="n">use_gaps</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

      <span class="n">Alignment_Name</span> <span class="n">Alignment_Length</span> <span class="n">Alignment_Percent_GC</span> <span class="n">Percent_Degenerate</span> <span class="n">Percent_N</span> <span class="n">Percent_Gap</span> <span class="n">Species_GC_Mean</span> <span class="n">Species_GC_StdDev</span> <span class="n">Percent_Nonbase</span> <span class="n">Percent_Invariant</span> <span class="n">Percent_Singleton</span> <span class="n">Percent_Parsimony_Informative</span>
<span class="m">1</span> <span class="n">Dummy_Alignment.fa</span>               <span class="m">13</span>            <span class="m">0.1470588</span>         <span class="m">0.03846154</span> <span class="m">0.1384615</span>  <span class="m">0.03846154</span>       <span class="m">0.1686147</span>         <span class="m">0.1825288</span>       <span class="m">0.1538462</span>         <span class="m">0.2307692</span>         <span class="m">0.1538462</span>                     <span class="m">0.3846154</span>
  <span class="n">Percent_Biallelic</span> <span class="n">Percent_Triallelic</span> <span class="n">Percent_Quadallelic</span> <span class="n">Percent_Pentallelic</span>
<span class="m">1</span>         <span class="m">0.2307692</span>         <span class="m">0.07692308</span>           <span class="m">0.1538462</span>                   <span class="m">0</span>
</pre></div>
</div>
<p><strong>Dummy Alignment</strong></p>
<a class="reference internal image-reference" href="../../_images/Dummy_Align.png"><img alt="../../_images/Dummy_Align.png" src="../../_images/Dummy_Align.png" style="width: 600px;" /></a>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="getAlignmentPatterns.html" class="btn btn-neutral float-right" title="getAlignmentPatterns" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="getAlignmentComposition.html" class="btn btn-neutral float-left" title="getAlignmentComposition" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Robert Literman

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>